File: interleaving_string.cpp
#include "header.h"

bool isInterleave(string s1, string s2, string s3) {
    int s1_sz = s1.size();
    int s2_sz = s2.size();
    int s3_sz = s3.size();
    if(s1_sz+s2_sz != s3_sz) return false;
    bool arr[s3_sz+1][s1_sz+1];
    arr[1][1] = s3[0]==s1[0];
    arr[1][0] = s3[0]==s2[0];
    for(int i=2; i<=s3_sz; i++){
        bool x = false;
        for(int j=0;j<=std::min(i,s1_sz);j++){
            if(i!=j && s3[i-1] == s2[i-j-1] && arr[i-1][j])
                x = true;
            else if(j>0 && s3[i-1]==s1[j-1] && arr[i-1][j-1])
                x = true;
            else
                x = false;
            arr[i][j] = x;
        }
    }
    return arr[s3_sz][s1_sz];
}

int main(){
//    string s1("aacaac");
//    string s2("aacaaeaac");
//    string s3("aacaaeaaeaacaac");
    string s1("c");
    string s2("");
    string s3("a");
    string s4("aadbbbaccc");
    printf("%s\n", isInterleave(s1,s2,s3)?"true":"false");
    printf("%s\n", isInterleave(s1,s2,s4)?"true":"false");
    return 0;
}

File: interleaving_string.py
class Solution:
    # @return a boolean
    def isInterleave(self, s1, s2, s3):
        if len(s1) == 0:
            return s2 == s3
        if len(s2) == 0:
            return s1 == s3
        s1_len = len(s1)
        s2_len = len(s2)
        s3_len = len(s3)
        if s1_len+s2_len != s3_len:
            return False
        arr = [[False for i in range(s3_len+1)] for j in range(s1_len+1)]
        arr[0][1] = True if s2[0]==s3[0] else False
        arr[1][1] = True if s1[0]==s3[0] else False
        for i in range(2,s3_len+1,1):
            for j in range(0,min(i,s1_len)+1,1):
                if i-j>s2_len:
                    continue
                if j>0 and s3[i-1]==s1[j-1] and arr[j-1][i-1]:
                    arr[j][i] = True
                elif i!=j and s3[i-1]==s2[i-j-1] and arr[j][i-1]:
                    arr[j][i] = True
        return arr[s1_len][s3_len]

print Solution().isInterleave("aabcc", "dbbca", "aadbbcbcac")
print Solution().isInterleave("aabcc", "dbbca", "aadbbbaccc")
